File: JavaScript/652-Find Duplicate Subtrees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode[]}
 */
var findDuplicateSubtrees = function(root) {
    let hmap = new Map();
    let res = [];
    
    let dfs = (root) =>{
        if (!root) return '#';
        let serial = root.val.toString() + "," + dfs(root.left) + "," + dfs(root.right);
        hmap.set(serial, (hmap.get(serial)||0) + 1);
        if (hmap.get(serial) === 2){
            res.push(root);
        }
        return serial;
    }
    dfs(root);
    return res;
};

File: 652-Find Duplicate Subtrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        subID = dict()
        count = collections.defaultdict(int)
        res = []
        def dfs(node):
            if not node:
                return 0
            triplet = (dfs(node.left), node.val, dfs(node.right))
            if triplet not in subID:
                subID[triplet] = len(subID) + 1
            ID = subID[triplet]
            count[ID] += 1
            if count[ID] == 2:
                res.append(node)
            return ID
        
        dfs(root)
        return res